### PR TITLE
string accessor: move back to ByteBuffer for Android

### PR DESCRIPTION
**Pull request by junr03**
Description: in general java String usage accross the JNI layer is precarious. Reverting usage back to ByteBuffer.
Risk Level: low 
Testing: tested with modified dispatcher code in example app that accessed the installed string accessor in MainActivity.kt

### CODE REVIEW

File: examples/kotlin/hello_world/MainActivity.kt [hunk 1]
```diff
@@ -48,7 +49,7 @@ class MainActivity : Activity() {
       .addPlatformFilter { DemoFilter() }
       .addPlatformFilter { BufferDemoFilter() }
       .addPlatformFilter { AsyncDemoFilter() }
-      .addStringAccessor("demo-accessor", { "PlatformString" })
+      .addStringAccessor("demo-accessor", { ByteBuffer.wrap("PlatformString".toByteArray(Charsets.UTF_8)) })
```
Comment (buildbreaker) [inline examples/kotlin/hello_world/MainActivity.kt:52]: What's fun is that this doesn't catch the existing issue... :(

**Review by buildbreaker — DISMISSED** [review]
Do we have an issue for patching up the test scenarios here @junr03 ?

If not, I can create one for us

File: library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt [hunk 2]
```diff
@@ -162,7 +163,7 @@ open class EngineBuilder(
    *
    * @return this builder.
    */
-  fun addStringAccessor(name: String, accessor: () -> String): EngineBuilder {
+  fun addStringAccessor(name: String, accessor: () -> ByteBuffer): EngineBuilder {
```
Comment (goaway) [inline library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt:166]: IMO the public interface should still return a String. We can map it to ByteBuffer behind the scenes.

Comment (goaway) [reply library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt:166]: Also why not use the ByteArray directly? We can't predict the lifetime of its usage, so we have to do a copy anyways - might as well just do it via the critical array copy that we already have.

Comment (buildbreaker) [reply library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt:166]: @goaway I feel like we can revisit our ByteBuffer versus ByteArray conversation again. Though I do agree if we can have a String interface but internally, it just transforms to a ByteBuffer, that would be nice

Comment (goaway) [reply library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt:166]: My point though is that we're converting to a byte array and then wrapping with a byte buffer when we already support passing byte array directly.

File: examples/kotlin/hello_world/MainActivity.kt (hunk 1 above)
Comment (junr03) [reply examples/kotlin/hello_world/MainActivity.kt]: Yeah, this only tests the install path, not the usage path.

When I originally implemented the change I modified the Http::Dispatcher to access the string, and obviously also tested using our internal filters. I failed to do that when I modified to String before :/, which is why I didn't catch the issues I am fixing now.

We could create a filter in Envoy Mobile that tests the example accessors. I will do a subsequent PR for that. We will just not install the filter in our private build.

File: library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt (hunk 2 above)
Comment (junr03) [reply library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt]: @buildbreaker @goaway updated. Pretty pleased with how it came out. Easy and ergonomic for the user, and consistent with prior JNI experience under the hood. Lmk what you think.

File: library/java/io/envoyproxy/envoymobile/engine/types/EnvoyStringAccessor.java [hunk 3]
```diff
@@ -1,5 +1,7 @@
 package io.envoyproxy.envoymobile.engine.types;
 
+import java.nio.ByteBuffer;
```
Comment (goaway) [inline library/java/io/envoyproxy/envoymobile/engine/types/EnvoyStringAccessor.java:3]: Looks like this can be dropped. :)

**Review by goaway — APPROVED** [review]
Looks good - see minor comment above. Thanks for updating!

File: examples/kotlin/hello_world/MainActivity.kt (hunk 1 above)
Comment (buildbreaker) [reply examples/kotlin/hello_world/MainActivity.kt]: https://github.com/envoyproxy/envoy-mobile/issues/1317!

Comment (junr03) [reply examples/kotlin/hello_world/MainActivity.kt]: @buildbreaker already have the test code written. Will add in a subsequent PR. Thanks for creating the issue.

**Review by buildbreaker — APPROVED** [review]

**Pull request merged**